common/src/test/java/io/netty/util/RecyclerTest.java

```diff
@@ -81,6 +81,38 @@ public void testMultipleRecycle() {
         object.recycle();
     }
 
+    @Test
+    public void testMultipleRecycleAtDifferentThread() {
+        Recycler<HandledObject> recycler = newRecycler(1024);
+        final HandledObject object = recycler.get();
+        final Thread thread1 = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                object.recycle();
+            }
+        });
+        thread1.start();
+        try {
+            thread1.join();
+        } catch (InterruptedException e) {
+        }
+
+        final Thread thread2 = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    object.recycle();
+                } catch (IllegalStateException e) {
+                }
```

Comment:
store the `IllegalStateException` in an `AtomicReference` and assert it after `thread2.join()` returns.